Generic StyleFactory to support span style for compose compatibility

folab33 opened the pull request.

markusressel commented:
Hey @folab33 ! :nerd_face: 
Looks like you put a lot of effort into this one :astonished: 
Before I take a deep dive into the code, could you give me a short description of what you did? :thinking:

folab33 commented:
> Hey @folab33 ! 🤓 Looks like you put a lot of effort into this one 😲 Before I take a deep dive into the code, could you give me a short description of what you did? 🤔

Jetpack compose doesn't work with SpannedString but with AnnotatedString. I made StyleFactory generic for it to work with other style like SpanStyle which will be use to produce annotated string in compose.

markusressel commented:
Dies this still work with SpannedString though? Or is it compose only?

folab33 commented:
It works with both SpannedString and compose. I modified SpannableHighlighter to make it work with the existing implementation and I added a new class AnnotatedStringHighlighter to produces annotated string that can be used in compose.

markusressel commented on app/src/main/AndroidManifest.xml:
```diff
@@ -9,7 +9,8 @@
         android:theme="@style/AppTheme">
-        <activity android:name=".MainActivity">
+        <activity android:name=".MainActivity"
+            android:exported="true">
```
Is this necessary?

folab33 replied:
Yes, apps targeting android 12 and above must declare this in activity manifest

markusressel commented on build.gradle:
```diff
@@ -8,13 +8,13 @@ buildscript {
 
-        minSdkVersion = 16
-        versionName = "3.1.0"
+        minSdkVersion = 21
```
I am fine with upping the minSdk, but this has to become a major release bump then. just make 4.0.0 right away.

markusressel reviewed (changes requested):
Looks good otherwise!

I am not too fond of the fact that Compose and legacy ColorSchemes have to be maintained separately now, but I also dont really see a better way...

markusressel commented:
I will have to make some tests on my device before merging, but looks good :+1: 
Thx!
